[#14] can-protocol-adapter initial version

RaviprasadBS opened the pull request:
#14 can-protocol-adapter initial version

RaviprasadBS committed:
[#14] can-protocol-adapter initial version

SebastianSchildt commented:
Thanks a lot. Not looked at the code yet, but I see pre-commit complains about some whitespace-relaled linting issues. If you install pre-commit as described here:  https://github.com/eclipse-kuksa/kuksa-incubation#pre-commit-set-up

It will always check and  - if needed - modify added files when you do "commit". 

To fix it after the fact, you can install pre-commit and  eitherrun

`pre-commit run --all-files`

and then _only_ add the files it modified in your folder, or, alternatively, check CI output for all files it complained about (https://github.com/eclipse-kuksa/kuksa-incubation/actions/runs/12904716321/job/35984188927?pr=18) and make a trivial edit (e.g. extra space or newline) to each file, so you can git add them. Then when you commit, pre-commit will run only on your added files, complain and fix them, You can then git add again and sucessfully commit

(Probably there are more ways, but those work for me)

RaviprasadBS committed:
[#14] Fixed pre-commit issues

RaviprasadBS commented:
Fixed the pre-commit errors and pushed the changes to branch, Thanks

lukasmittag commented on can-protocol-adapter/build.sh:
```diff
@@ -0,0 +1,67 @@
+
+# Run cargo clippy with -D warnings
+echo "Running cargo clippy..."
+cargo clippy -- -D warnings 2>&1 | tee $CLIPPY_LOG
```
https://github.com/eclipse-kuksa/kuksa-databroker/blob/86ed421573f320f6220e336b5962ec03a86f83a2/.github/workflows/kuksa_databroker_build.yml#L69C1-L69C67

in databroker build we use -w warnings as well

RaviprasadBS replied:
Updated in the commit id 75b9369.

lukasmittag commented on can-protocol-adapter/doc/configuration.md:
```diff
@@ -0,0 +1,112 @@
+The CAN ID used for receiving messages.
+
+### socket_can_type:
+Specifies the socket type for CAN communication, set to "SOCK_DGRAM", which is typical for datagram-based communication.
```
maybe here mention other possible options since some people may not be this familiar with socket can types. Or just link to the rust socket documentation there it should be mentioned :)

RaviprasadBS replied:
Updated in the commit id 75b9369

lukasmittag commented on can-protocol-adapter/doc/configuration.md:
```diff
@@ -0,0 +1,112 @@
+     "interval_ms": 500,
+     "dbc_signal_name": "S01PID0D_VehicleSpeed",
+     "vss_signal": {
+     "signal_name": "Vehicle.CAN.Speed",
```
better to use would be `Vehicle.Speed` right?

RaviprasadBS replied:
Yes .. Its the typo we have used Vehicle.OBD.Speed signal. I have updated the same in commit id 75b9369.

lukasmittag commented on can-protocol-adapter/doc/v0.1.0-initial-version.md:
```diff
@@ -0,0 +1,17 @@
+
+- JSON based user configuration for CAN interfaces, requests/responses and VSS to PID mapping.
+- Supports socket-can and ISOTP (ISO 15765-2) standards.
+- CAN frame decoding using use-defined DBC files.
```
```suggestion
- CAN frame decoding using user-defined DBC files.

```

RaviprasadBS replied:
Updated in the commit id 75b9369.

lukasmittag commented on can-protocol-adapter/lib/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+########################################################################
```
You could opt here to include databroker as a submodule. Thats what we did in the `kuksa-python-sdk` because otherwise this needs a lot of maintenance effort. However I think for now it is okay since we plan to relase the rust lib as a crate in the near future and the that this component lives in kuksa-incubation.

RaviprasadBS replied:
Yes We have used databroker as a submodule now. Updated in the commit id afb7ee3.

lukasmittag commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+    let broker_port = adapter_config.general_config.broker_port.clone();
+
+    // Create a new Feeder instance and connect to the data broker.
+    let mut feeder = Feeder::new(broker_ip.clone(), broker_port.clone());
```
Just some naming guideline since we moved from the old github organization we refer to a `Feeder` as a `Provider` would be nice if you could rename :)

lukasmittag commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+
+    // Register the user defined datapoints with the data broker.
+    let datapoints = adapter_utils::datapoints_from_config(&adapter_config);
+    match feeder.register_datapoints(datapoints).await {
```
this is deprecated you could continue to use it, but then you should mention in your documentation that you need to start the databroker with `enable-databroker-v1`. Better option would be to check if the datapoints are there with `GetRequest` where you specify that you want the metadata of a signal.

RaviprasadBS replied:
Updated the Rreadme.MD documentation as suggested.

lukasmittag commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+
+    // Create shared resources using Arc and Mutex.
+    let shared_socket = Arc::new(Mutex::new(socket));
+    let shared_feeder = Arc::new(Mutex::new(feeder));
```
```suggestion
    let shared_provider = Arc::new(Mutex::new(provider));
```

RaviprasadBS replied:
Updated in the commit id 75b9369.

lukasmittag commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+    let receive_task_handle = tokio::spawn({
+        let socket_instance = Arc::clone(&shared_socket);
+        let adapter_config = Arc::clone(&adapter_config);
+        let feeder_instance = Arc::clone(&shared_feeder);
```
```suggestion
        let provider_instance = Arc::clone(&shared_provider);
```

RaviprasadBS replied:
Updated in the commit id 75b9369.

lukasmittag commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+            comm::receive_can_data(
+                socket_instance,
+                adapter_config,
+                feeder_instance,
```
```suggestion
                provider_instance,
```

RaviprasadBS replied:
Updated in the commit id 75b9369.

lukasmittag commented on can-protocol-adapter/src/databroker-proto/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+#********************************************************************************
```
see comment about submodule above

RaviprasadBS replied:
Updated in the commit id 75b9369.

RaviprasadBS replied:
@lukasmittag  Thanks for review comments. I have addressed all the review comments and committed the changes to my branch. Please revert back to me if there is any queries.

lukasmittag reviewed (commented):
Some comments I have to the code. Mainly structuring and maintenance thingys.

RaviprasadBS committed:
submodule databroker is added

RaviprasadBS committed:
removed unused files

sophokles73 commented on can-protocol-adapter/README.md:
```diff
@@ -0,0 +1,100 @@
+
+## The basic functionality works as follows:
+
+The application connects to the SocketCAN interface and sends all CAN requests, as specified in the PID table of the config file, based on a defined interval. The responses are read, and the raw CAN data is parsed according to the DBC file signals. The CAN signals are then translated into VSS data points based on the VSS and DBC signal mappings described in the config file. The respective VSS dat ...
```
In the diagram the _Application_ seems to be the component that interacts with the Kuksa Databroker. In your text, _application_ seems to refer to the _CAN Protocol Adapter_, though. Or am I mistaken? IMHO you should try to be consistent in the terminology in order to not confuse the reader.

RaviprasadBS replied:
Corrected the document. Commit id : 6a421db

RaviprasadBS committed:
[#14] Addressed review comments

RaviprasadBS committed:
Update README.md

sophokles73 commented on can-protocol-adapter/src/can/comm.rs:
```diff
@@ -0,0 +1,204 @@
+        .map(|entry| Instant::now() + Duration::from_millis(entry.interval_ms as u64))
+        .collect();
+
+    loop {
```
Have you considered using an existing task scheduler crate (e.g. https://crates.io/crates/tokio-task-scheduler) instead of implementing this yourself?

RaviprasadBS replied:
Thanks for the suggestion. We were not aware of the "tokio-task-scheduler" crate. we will consider it for the next update.

sophokles73 commented on can-protocol-adapter/src/main.rs:
```diff
@@ -0,0 +1,160 @@
+            decoder
+        }
+        Err(_e) => {
+            panic!("Critical error: Could not parse DBC file. Exiting.");
```
In the rest of the main function your error handling strategy seems to be to return an `Err` instead of panicking ...

RaviprasadBS replied:
Updated the code to return Err instead of panicking.

sophokles73 commented on can-protocol-adapter/src/utils/adapter_config.rs:
```diff
@@ -0,0 +1,86 @@
+    u32::from_str_radix(s.trim_start_matches("0x"), 16).map_err(serde::de::Error::custom)
+}
+
+fn from_hex_string_to_bytes<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
```
you might want to consider using [JSON5](https://json5.org/) for the config file, which already supports using hex numbers (and much  more) in the JSON. It can also be used with [serde]( https://crates.io/crates/serde_json5)

RaviprasadBS replied:
Thank you for the helpful suggestions. We will consider updating to JSON5 in the next release and will include this feature in the 'Planned Enhancements' section of the documentation.

sophokles73 commented on can-protocol-adapter/src/utils/adapter_utils.rs:
```diff
@@ -0,0 +1,102 @@
+    Ok(())
+}
+
+pub fn datapoints_from_config(adapter_config: &AdapterConfig) -> HashMap<String, v1::Metadata> {
```
In order to increase encapsulation and cohesion, you might want to move this function to the `kuksa_provider::provider` module and adapt `Provider::register_datapoints` to accept an `AdapterConfig` instead ...

RaviprasadBS replied:
As suggested, moved "datapoints_from_config" to kuksa_provider::provider module and made necessary changes to Provider::register_datapoints function.

sophokles73 commented on can-protocol-adapter/src/can/decoder.rs:
```diff
@@ -0,0 +1,112 @@
+                    signal.offset()
+                );
+            }
+        }
```
you might want to do this only if log level is <= TRACE ...

RaviprasadBS replied:
Added a condition to execute only if the log level is <= TRACE.

sophokles73 commented on can-protocol-adapter/src/can/decoder.rs:
```diff
@@ -0,0 +1,112 @@
+use std::mem;
+
+pub struct Decoder {
+    dbc: Option<DBC>,
```
in what case would this be None? If the DBC file cannot be opened/parsed, then the `new` function returns an error, doesn't it?

RaviprasadBS replied:
Yes, that's correct. If the DBC file cannot be opened or parsed, the new function returns an error, not None. I have removed the dbc: Option<DBC> as i am using HashMap to map signal name to signal def as part of the new.

sophokles73 commented on can-protocol-adapter/src/can/decoder.rs:
```diff
@@ -0,0 +1,112 @@
+            .iter()
+            .flat_map(|message| message.signals())
+            .find(|signal| signal.name() == signal_name)
+            .ok_or_else(|| format!("Signal '{}' not found in DBC", signal_name))?;
```
doing this for each and every received frame seems highly inefficient, given that the DBC content does not change during runtime. How about creating a HashMap mapping signal names to signal definitions as part of the `new` function?

RaviprasadBS replied:
Yes, that's correct. I have created HashMap mapping signal names to signal definitions  as part of new and removed the redundant processing during runtime.

sophokles73 commented on can-protocol-adapter/src/can/comm.rs:
```diff
@@ -0,0 +1,204 @@
+                //Release the socket lock.
+                drop(socket_lock);
+                debug!("CAN Tx: PID: {:X?}", entry.request_pid);
+                if let Err(err) = pid_tx.send(i).await {
+                    error!("Error sending request info: {}", err);
+                }
+                // Wait for a response from the receive_can_data function , Max timeout 1 second
+                match time::timeout(Duration::from_secs(1), pid_rx.recv()).await {
```
My understanding of OBDII is that you send out a request on the CAN bus and then one or more ECUs send a response with the requested value. However, there is no real correlation between a response and request, i.e. you cannot determine the particular request that a response is the answer to. Based on that, I do not see why you would want/need to synchronize the sender and receiver tasks in this way. FMPOV sending the requests is fire and forget, i.e. the requests merely _trigger_ the periodic reporting of current values. All of the responses can then simply be processed as they arrive, or am I mistaken?

sophokles73 commented on can-protocol-adapter/lib/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+[workspace]
+resolver = "2"
+
+members = [
```
FMPOV it is quite confusing that you have _embedded_ this multi crate workspace within the top level crate (_can-protocol-adapter_).
The folder structure would be much easier to understand (and manage) if you would instead have just a single top level workspace which has all of the crates as members, i.e. _can-protocol-adapter_, _common_ and _kuksa_.

RaviprasadBS replied:
Yes, you're correct. I have now refactored the structure and removed the common and kuksa lib sources from the folder structure, as they are now used from submodules.

RaviprasadBS committed:
[#14] Addressed review comments

sophokles73 reviewed (approved):
LGTM

RaviprasadBS committed:
Update v0.1.0-initial-version.md

RaviprasadBS committed:
[#14] Add can simulation script

RaviprasadBS committed:
Update README.md

RaviprasadBS committed:
[#14] Fixed pre-commit issues

RaviprasadBS commented:
@SebastianSchildt @lukasmittag @sophokles73  Thank you for the review comments. I have addressed all of them and committed the changes to my branch. Additionally, I have added a CAN simulator file, which can be used to mock CAN communication.

SebastianSchildt commented:
Thank you for the modifications @RaviprasadBS , looks goodand 99.9% done for initial merge I think , however  there are some minor linting issues in Readme.md (https://github.com/eclipse-kuksa/kuksa-incubation/actions/runs/13157739283/job/36721958113?pr=18 some extra spaces I guess)

RaviprasadBS committed:
[#14] Fix linting issues in Readme.md

RaviprasadBS commented:
@SebastianSchildt  I have corrected the linting issue in Readme.md file and committed the changes. Thanks

SebastianSchildt reviewed (approved):
Looking good. Let's merge this. Looking forward to new PRs with new features/improvements!
